lib/Mojolicious/Plugin/OpenAPI/Cors.pm

```diff
@@ -48,63 +169,187 @@ Mojolicious::Plugin::OpenAPI::Cors - OpenAPI plugin for Cross-Origin Resource Sh
+    # or just let the auto-renderer handle it.
+    $c->res->headers->header("Access-Control-Allow-Origin" => $params->{origin});
+
+    # Set Preflighted response headers, instead of using the default
+    if ($params->{type} eq "preflighted") {
+      $c->res->headers->header("Access-Control-Allow-Headers" => "X-Whatever, X-Something");
+      $c->res->headers->header("Access-Control-Allow-Methods" => "POST, GET, OPTIONS");
+      $c->res->headers->header("Access-Control-Max-Age" => 86400);
     }
+
+    # Return undef on success.
+    return undef;
   }
 
 =head1 DESCRIPTION
 
-L<Mojolicious::Plugin::OpenAPI::Cors> is a plugin for accepting Simple
-Cross-Origin Resource Sharing requests, by looking at the "Origin" header. See
+L<Mojolicious::Plugin::OpenAPI::Cors> is a plugin for accepting Preflighted or
+Simple Cross-Origin Resource Sharing requests. See
 L<https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS> for more details.
 
 This plugin is loaded by default by L<Mojolicious::Plugin::OpenAPI>.
 
-Note that this plugin currently EXPERIMENTAL! Please let me know if you have
-any feedback.
+Note that this plugin currently EXPERIMENTAL! Please comment on
+L<https://github.com/jhthorsen/mojolicious-plugin-openapi/pull/102> if
+you have any feedback or create a new issue.
+
+=head1 STASH VARIABLES
+
+The following "stash variables" can be set in L<Mojolicious/defaults>,
+L<Mojolicious::Routes::Route/to> or L<Mojolicious::Controller/stash>.
+
+=head2 openapi_cors_allow_origins
+
+This variable should hold an array-ref of regexes that will be matched against
+the "Origin" header in case the default
+L</openapi_cors_default_exchange_callback> is used. Examples:
+
+  $app->defaults(openapi_cors_allow_origins => [qr{^https?://whatever.example.com}]);
+  $c->stash(openapi_cors_allow_origins => [qr{^https?://whatever.example.com}]);
+
+=head2 openapi_cors_default_exchange_callback
+
+Instead of using the default C<$callback> provided by this module for
+L</openapi.cors_exchange>, you can set a global value.
+
+=head2 openapi_cors_default_max_age
+
+Holds the default value for the "Access-Control-Max-Age" response header
+set by L</openapi.cors_preflighted>. Examples:
+
+  $app->defaults(openapi_cors_default_max_age => 3600);
+  $c->stash(openapi_cors_default_max_age => 3600);
 
 =head1 HELPERS
 
-=head2 openapi.cors_simple
+=head2 openapi.cors_exchange
+
+  $c = $c->openapi->cors_exchange($callback);
+  $c = $c->openapi->cors_exchange("MyApp::cors_validator");
+  $c = $c->openapi->cors_exchange("_some_controller_method");
+  $c = $c->openapi->cors_exchange(sub { ... });
+  $c = $c->openapi->cors_exchange;
+
+Used to validate either a simple CORS request, preflighted CORS request or a
+real request. It will be called as soon as the "Origin" request header is seen.
+
+The C<$callback> will be called with the following arguments:
 
-  $c = $c->openapi->cors_simple($method);
+  my $error = $callback->($c, {
```

Comment:
As these values are available via the controller, I don't think I would have passed this hash to the user. That said, I do understand the convenience. Maybe document "as a convenience"?